### PR TITLE
Expand rpc_test to include 0, 1, and 2 hop virtual funding

### PR DESCRIPTION
**Pull request by NiloCK**
PR replaces `executeRpcTest` with `executeNRpcTest`, and calls it with 2, 3 and 4 clients.

Recommending to read the new function top-to-bottom (diff w/ removed function hard to parse).

### CODE REVIEW

File: client_test/rpc_test.go [hunk 1]
```diff
@@ -44,143 +46,244 @@ func createLogger(logDestination *os.File, clientName, rpcRole string) zerolog.L
...
-	checkNotifications(t, requiredVirtualNotifs, optionalVirtualNotifs, aliceVirtualNotifs, defaultTimeout)
 
-	checkNotifications(t, requiredVirtualNotifs, optionalVirtualNotifs, bobVirtualNotifs, defaultTimeout)
+	aliceLedgerNotifs := aliceClient.LedgerChannelUpdatesChan(ledgerChannels[0].ChannelId)
+	expectedAliceLedgerNotifs := createLedgerStory(
+		aliceLedger.ChannelId, alice.Address(), actors[1].Address(), // actor[1] is the first intermediary - can be Bob if n=2 (0-hop)
+		[]channelStatusShorthand{
+			{100, 100, query.Proposed},
+			{100, 100, query.Open},
+			{0, 100, query.Open},
```
Comment (kerzhner) [inline client_test/rpc_test.go:238]: I don't think it's the case, but is this the order we expect? I don't think we check the order of notification in `checkNotifications`, but this is still a bit confusing.

File: client_test/rpc_test.go [hunk 2]
```diff
@@ -44,143 +46,244 @@ func createLogger(logDestination *os.File, clientName, rpcRole string) zerolog.L
...
+			{99, 101, query.Open},
+			{99, 101, query.Complete},
+		},
+	)
+	if n != 2 { // bob does not trigger a ledger-channel close if n=2 - alice does
+		expectedBobLedgerNotifs = append(expectedBobLedgerNotifs,
```
Comment (kerzhner) [inline client_test/rpc_test.go:258]: Not a part of this work, but it's a bit confusing that Closing notification is only sent to the participant that initiated the close. Do we know why this is the case?

File: client_test/rpc_test.go [hunk 3]
```diff
@@ -44,143 +46,244 @@ func createLogger(logDestination *os.File, clientName, rpcRole string) zerolog.L
...
+
+	bobLedgerNotifs := bobClient.LedgerChannelUpdatesChan(bobLedger.ChannelId)
+	expectedBobLedgerNotifs := createLedgerStory(
+		bobLedger.ChannelId, actors[n-2].Address(), bob.Address(),
+		[]channelStatusShorthand{
+			{100, 100, query.Proposed},
+			{100, 100, query.Open},
+			{0, 100, query.Open},
+			{99, 101, query.Open},
+			{99, 101, query.Complete},
```
Comment (kerzhner) [inline client_test/rpc_test.go:254]: It's really helpful to see all of the expected notifications listed out succinctly.

**Review by kerzhner — APPROVED** [review]
Nicely done!

File: client_test/rpc_test.go (hunk 1 above)
Comment (NiloCK) [reply client_test/rpc_test.go:238]: This is the correct order - the shorthand is not a complete description, so it's not as obvious as it could be.

- A:100, I:100
- A:0, I:100, G(vAB): 100  (Alice forwards 100 to the guarantee)
- A: 99, I: 101 (After vAB closes, A and I's balances adjust)

Comment (NiloCK) [reply client_test/rpc_test.go:238]: ... But in any case the test itself is order insensitive for this data.

Comment (NiloCK) [reply client_test/rpc_test.go]: fc952786

File: client_test/rpc_test.go [hunk 4]
```diff
@@ -44,143 +46,244 @@ func createLogger(logDestination *os.File, clientName, rpcRole string) zerolog.L
...
+	for i := 0; i < n-1; i++ {
+		outcome := simpleOutcome(actors[i].Address(), actors[i+1].Address(), 100, 100)
+		ledgerChannels[i] = clients[i].CreateLedger(actors[i+1].Address(), 100, outcome)
+
+		if !directfund.IsDirectFundObjective(ledgerChannels[i].Id) {
+			t.Errorf("expected direct fund objective, got %s", ledgerChannels[i].Id)
+		}
+
+		// wait for the ledger channel to be ready
+		<-clients[i].ObjectiveCompleteChan(ledgerChannels[i].Id)
```
[the remark this answers is not in the document]
Comment (NiloCK) [reply client_test/rpc_test.go]: 535d6bff

File: client_test/rpc_test.go (hunk 2 above)
Comment (NiloCK) [reply client_test/rpc_test.go:260]: I do not. Feels like a bug - I can follow up on it next week.

File: client_test/rpc_test.go [hunk 5]
```diff
@@ -287,10 +390,15 @@ func marshalToJson[T channelInfo](t *testing.T, info T) string {
...
+//
+// required specifies the notifications that must be received. checkNotifications will fail
+// if any of these notifications are not received.
+//
+// optional specifies notifications that may be received. checkNotifications will not fail
+// if any of these notifications are not received.
+//
 // If a notification is received that is neither in required or optional, checkNotifications will fail.
-func checkNotifications[T channelInfo](t *testing.T, required []T, optional []T, notifChan <-chan T, timeout time.Duration) {
+func checkNotifications[T channelInfo](t *testing.T, client string, required []T, optional []T, notifChan <-chan T, timeout time.Duration) {
```
[the remark this answers is not in the document]
Comment (NiloCK) [reply client_test/rpc_test.go:403]: Hmm...
